src/SharedStream.mo

```diff
@@ -0,0 +1,380 @@
+    source : StreamSource;
+    var nextItemId : Nat;
+    var receiver : ?StreamReceiver<T>;
+  };
+
+  type StableStreamInfo = {
+    source : StreamSource;
+    nextItemId : Nat;
+    active : Bool;
+  };
+
+  public type ManagerStableData = (Vec.Vector<StableStreamInfo>, Vec.Vector<(Principal, ?Nat)>);
+  public func defaultManagerStableData() : ManagerStableData = (Vec.new(), Vec.new());
+
+  func requireOk<T>(res : R.Result<T, Any>) : T {
+    switch (res) {
+      case (#ok ok) ok;
+      case (_) Debug.trap("Required result is #err");
+    };
+  };
+
+  public class StreamsManager<T>(
+    initialSourceCanisters : [Principal],
+    itemCallback : (streamId : Nat, sourceCanisterIndex : ?Nat, item : T, index : Nat) -> Any,
+  ) {
+
+    // info about each issued stream id is preserved here forever. Index is a stream ID
+    let streams_ : Vec.Vector<StreamInfo<T>> = Vec.new();
+    // a list of registered source canister principals
+    let sourceCanisters_ : Vec.Vector<Principal> = Vec.fromArray(initialSourceCanisters);
+    // a list of source canister current stream id-s, can be mapped to 'sourceCanisters' by indices
+    let sourceCanisterStreamIds_ : Vec.Vector<?Nat> = Vec.init<?Nat>(initialSourceCanisters.size(), null);
+
+    public func sourceCanisters() : Vec.Vector<Principal> = sourceCanisters_;
+
+    public func getStream(id : Nat) : ?StreamInfo<T> = Vec.getOpt(streams_, id);
+
+    public func getNextStreamId() : Nat = Vec.size(streams_);
+
+    public func issueStreamId(source : StreamSource) : R.Result<Nat, { #NotRegistered }> {
+      let id = Vec.size(streams_);
+      switch (source) {
+        case (#canister p) switch (Vec.indexOf(p, sourceCanisters_, Principal.equal)) {
+          case (null) return #err(#NotRegistered);
+          case (?srcIndex) {
+            switch (Vec.get(sourceCanisterStreamIds_, srcIndex)) {
+              case (?oldStreamId) Vec.get(streams_, oldStreamId).receiver := null;
+              case (_) {};
+            };
+            Vec.put(sourceCanisterStreamIds_, srcIndex, ?id);
+          };
+        };
+        case (#internal) {};
+      };
+      Vec.add(
+        streams_,
+        {
+          source = source;
+          var nextItemId = 0;
+          var receiver = ?StreamReceiver<T>(id, streamCallback, 0);
+        },
+      );
+      #ok id;
+    };
+
+    func streamCallback(streamId : Nat, item : T, index : Nat) {
+      let stream = Vec.get(streams_, streamId);
+      let sourceIndex : ?Nat = switch (stream.source) {
+        case (#canister p) Vec.indexOf(p, sourceCanisters_, Principal.equal);
+        case (_) null;
+      };
+      stream.nextItemId += 1;
+      ignore itemCallback(streamId, sourceIndex, item, index);
+    };
+
+    public func issueInternalStreamId() : Nat = requireOk(issueStreamId(#internal));
+
+    public func registerSourceCanister(p : Principal) : () {
+      Vec.add(sourceCanisters_, p);
+      Vec.add(sourceCanisterStreamIds_, null);
```

Comment:
Here we should check that we don't create duplicate entries (if p already exists in the Vector).

Consider using an AssocList.

Reply:
Done